pkg/apis/v1/dnsrecord_types.go

```diff
@@ -221,6 +221,78 @@ type DNSZoneCondition struct {
 	Message            string      `json:"message,omitempty"`
 }
 
+const (
+	TargetTypeHost = "HOST"
+	TargetTypeIP   = "IP"
+)
+
+type Target struct {
+	Cluster    string
+	TargetType string
+	Value      string
+}
+
+func (endpoint *Endpoint) GetAddress() (string, bool) {
+	if endpoint.SetIdentifier == "" || len(endpoint.Targets) == 0 {
+		return "", false
+	}
+
+	return string(endpoint.Targets[0]), true
+}
+
+func (endpoint *Endpoint) SetProviderSpecific(name, value string) {
+	var property *ProviderSpecificProperty
+
+	if endpoint.ProviderSpecific == nil {
+		endpoint.ProviderSpecific = ProviderSpecific{}
+	}
+
+	for _, pair := range endpoint.ProviderSpecific {
+		if pair.Name == name {
+			property = &pair
+		}
+	}
+
+	if property != nil {
+		property.Value = value
+		return
+	}
+
+	endpoint.ProviderSpecific = append(endpoint.ProviderSpecific, ProviderSpecificProperty{
+		Name:  name,
+		Value: value,
+	})
+}
+
+func (endpoint *Endpoint) GetProviderSpecific(name string) (string, bool) {
+	for _, property := range endpoint.ProviderSpecific {
+		if property.Name == name {
+			return property.Value, true
+		}
+	}
+
+	return "", false
+}
+
+func (endpoint *Endpoint) DeleteProviderSpecific(name string) bool {
+	if endpoint.ProviderSpecific == nil {
+		return false
+	}
+
+	deleted := false
+	providerSpecific := make(ProviderSpecific, 0, len(endpoint.ProviderSpecific))
+	for _, pair := range endpoint.ProviderSpecific {
+		if pair.Name == name {
+			deleted = true
+		} else {
+			providerSpecific = append(providerSpecific, pair)
+		}
+	}
+
+	endpoint.ProviderSpecific = providerSpecific
+	return deleted
+}
+
```

Comment:
I assume these were copied from kcp-glbc. These functions had duplicates and were a bit all over the place as i think two people added similar things. This was the reason i didn't just add them all as is form kcp-glbc in the first place. Just noting we should come back to these and make it consistent in how they work are named. e.g. There is already a GetProviderSpecificProperty function. Doesn't need to block a merge.